featurebyte/service/scd_table_validation.py

```diff
@@ -0,0 +1,206 @@
+"""
+SCDTableValidationService
+"""
+
+from __future__ import annotations
+
+import pandas as pd
+from sqlglot import expressions
+from sqlglot.expressions import select
+
+from featurebyte.enum import AggFunc, SpecialColumnName
+from featurebyte.exception import SCDTableValidationError
+from featurebyte.query_graph.node.schema import TableDetails
+from featurebyte.query_graph.sql.adapter import BaseAdapter
+from featurebyte.query_graph.sql.asat_helper import (
+    get_record_validity_condition,
+)
+from featurebyte.query_graph.sql.ast.literal import make_literal_value
+from featurebyte.query_graph.sql.common import (
+    get_qualified_column_identifier,
+    quoted_identifier,
+    sql_to_string,
+)
+from featurebyte.query_graph.sql.groupby_helper import GroupbyColumn, GroupbyKey, get_groupby_expr
+from featurebyte.query_graph.sql.materialisation import get_source_expr
+from featurebyte.schema.scd_table import SCDTableCreate
+from featurebyte.session.base import BaseSession
+
+COUNT_PER_NATURAL_KEY = "COUNT_PER_NATURAL_KEY"
+
+
+class SCDTableValidationService:
+    """
+    SCDTableValidationService class
+    """
+
+    async def validate_scd_table(
+        self,
+        session: BaseSession,
+        table_creation_payload: SCDTableCreate,
+    ) -> None:
+        """
+        Check that a table is a valid Slowly Changing Dimension (SCD) table based on the provided
+        table creation payload
+
+        Parameters
+        ----------
+        session: BaseSession
+            Session object
+        table_creation_payload: SCDTableCreate
+            SCD table creation payload
+
+        Raises
+        ------
+        SCDTableValidationError
+            If the table is not a proper SCD table
+        """
+        if table_creation_payload.natural_key_column is None:
+            return
+
+        natural_key_column = table_creation_payload.natural_key_column
+
+        # Check if there are multiple active records as of now. Only need to check if
+        # end_timestamp_column is present since otherwise with the inferred end timestamp, there
+        # will not be multiple active records.
+        if table_creation_payload.end_timestamp_column is not None:
+            query = self._get_active_record_counts_as_at_now(
+                session.adapter,
+                table_details=table_creation_payload.tabular_source.table_details,
+                effective_timestamp_column=table_creation_payload.effective_timestamp_column,
+                natural_key_column=natural_key_column,
+                end_timestamp_column=table_creation_payload.end_timestamp_column,
+            )
```

Comment:
If i understood the function correctly, it will only return the rows that has more than 1 active record, right?

Reply:
That is right. I think I should rename the method names to be clearer, e.g. `_get_invalid_active_records_as_at_now`.